src/sql/src/names.rs

```diff
@@ -632,55 +620,127 @@ impl AstInfo for Aug {
 
 /// The identifier for a schema.
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
-pub struct SchemaId(pub u64);
+#[serde(into = "SchemaIdJson")]
+#[serde(try_from = "SchemaIdJson")]
+pub enum SchemaId {
+    User(u64),
+    System(u64),
+}
+
+/// A type that specifies how we serialize and deserialize a [`SchemaId`].
+///
+/// Note: JSON maps require their keys to be strings, and we use [`SchemaId`] as the key
+/// in some maps, so we need to serialize them as plain strings, which is what this type
+/// does.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(transparent)]
+struct SchemaIdJson(String);
```

Comment:
Ideally this should belong in `storage.rs` or somewhere nearby, since this create and module don't normally care how things are stored on disk.

I have a slight preference (but only slight) to use 
```
#[serde(into = "BTreeMap<String, _>")]
#[serde(try_from = "BTreeMap<String, _>")]
```
where ever this is being used as a key if that means we can move this logic out of the `sql` create.

Reply:
This is only necessary for [dumping the catalog](https://github.com/parker-timmerman/materialize/blob/main/src/adapter/src/catalog.rs#L1350-L1357) to a string. In terms of location, I did try putting it in `storage.rs` but it kind of felt out of place there since its not really related to how we persist things to disk, since it's only used for this one function. Instead of these custom impls, something else I tried was the following:

```
pub fn dump(&self) -> String {
  let database_by_str: BTreeMap<String, _> = self.database_by_id.iter()
    .map(|(key, value)| (key.to_string(), value.debug_json()))
    .collect();
  serde_json::to_string(database_by_str).expect("serialization cannot fail")
}

impl Database {
    fn debug_json(&self) -> serde_json::Value {
        let schemas_by_str: BTreeMap<String, _> = self
            .schemas_by_id
            .iter()
            .map(|(key, value)| (key.to_string(), value))
            .collect();
        
        serde_json::json!({
            name: self.name,
            id: self.id,
            schemas_by_id: schemas_by_str,
            schemas_by_name: self.schemas_by_name,
            owner_id: self.owner_id,
        })
    }
}
```

Maybe that's a bit better?

Reply:
I slightly prefer your proposed approach compared to the current approach. The current approach forces us to always serialize `SchemaId` as a string while your proposed approach only serializes `SchemaId` as a string when it's required (in the `dump` function. I can imagine in the future if `SchemaId` is serialized in the stash then we would want to serialize it normally and not as a string.

This is all subjective though, so I'll leave the final decision up to you.

Reply:
Updated to use the `debug_json(...)` approach! Since we only need this "special" case serialization for a single method, I think putting the special behavior in that method makes the most sense